### PR TITLE
Added Ctrl-L to FSI

### PR DESCRIPTION
**Pull request by voronoipotato**
I added this feature to FSI. I also cleared up a bit of the code where it was using character codes instead of keys.

### CODE REVIEW

**Comment by dsyme** [comment]
@voronoipotato Could you jot down what manual testing you have done on this? Thanks

File: src/fsharp/fsi/console.fs [hunk 1]
```diff
@@ -377,7 +378,25 @@ type internal ReadLineConsole() =
...
+                    Console.Clear()
+                    Console.Write ( x.Prompt)
+                    Console.Write( input.ToString())
+                    anchor := Anchor.Top(x.Inset)   
+            let previous = history.Previous()
+            history.Next() |> ignore
+            if previous = "" then
+                setPrompt(true)
+            else
+                setPrompt(previous.[previous.Length - 2..previous.Length - 1] = ";;")
```
Comment (vasily-kirichenko) [inline src/fsharp/fsi/console.fs]: Is this always safe?

**Comment by voronoipotato** [comment]
Not a ton don, I just used it normally hitting ctrl-l while using fsi making sure it respected multi line, that it didn't nuke my history or my variables and functions already defined.

**Comment by voronoipotato** [comment]
I think Vasily is correct it will crash if you have a single character as your previous command. I'll fix that and address Carter's nits

**Comment by dsyme** [comment]
> Not a ton don, I just used it normally hitting ctrl-l while using fsi making sure it respected multi line, that it didn't nuke my history or my variables and functions already defined.

OK, could you run through all the key bindings implemented and make sure they all continue to work too, thanks

**Comment by cartermp** [comment]
@voronoipotato Have you had a chance to think about addressing the feedback?

**Comment by voronoipotato** [comment]
I have had a chance, sorry I've been busier than usual. I'll be able to work on it this weekend.

**Comment by voronoipotato** [comment]
I tested the other commands, and they behaved the same as before. The only things of note I found is that Ctrl-A highlights all for me, and crtl-f initiates find for me, but it also does this on the previous version. I changed the string as array access and comparison for EndsWith which is cleaner and simpler.

**Comment by cartermp** [comment]
@voronoipotato Have you had a chance to address feedback?

**Comment by voronoipotato** [comment]
Hmm I thought that's what my change covered. I thought I had but even if so I definitely flopped on proper communication, many apologies. I'll review it this evening and make sure it addresses all the feedback.  Thanks for the ping @cartermp .

File: src/fsharp/fsi/console.fs (hunk 1 above)
Comment (voronoipotato) [reply src/fsharp/fsi/console.fs]: fixed with EndsWith

File: src/fsharp/fsi/console.fs [hunk 2]
```diff
@@ -377,7 +378,25 @@ type internal ReadLineConsole() =
...
                     history.AddLast(line);
                 line;
-        
+        let clear() = 
+            current := input.Length
+            let setPrompt(prompt) = 
+                if prompt then //we only allow clearing if prompt is ">"
+                    Console.Clear()
+                    Console.Write ( x.Prompt)
+                    Console.Write( input.ToString())
```
[the remark this answers is not in the document]
Comment (voronoipotato) [reply src/fsharp/fsi/console.fs]: removed extra space

File: src/fsharp/fsi/console.fs [hunk 3]
```diff
@@ -377,7 +378,25 @@ type internal ReadLineConsole() =
                 if (line.Length > 0) then 
                     history.AddLast(line);
                 line;
-        
+        let clear() = 
+            current := input.Length
+            let setPrompt(prompt) = 
+                if prompt then //we only allow clearing if prompt is ">"
+                    Console.Clear()
+                    Console.Write ( x.Prompt)
```
[the remark this answers is not in the document]
Comment (voronoipotato) [reply src/fsharp/fsi/console.fs]: removed extra space

File: src/fsharp/fsi/console.fs [hunk 4]
```diff
@@ -377,7 +378,25 @@ type internal ReadLineConsole() =
                 if (line.Length > 0) then 
                     history.AddLast(line);
                 line;
-        
+        let clear() = 
+            current := input.Length
+            let setPrompt(prompt) = 
+                if prompt then //we only allow clearing if prompt is ">"
```
[the remark this answers is not in the document]
Comment (voronoipotato) [reply src/fsharp/fsi/console.fs]: added space after //, capitalized W

File: src/fsharp/fsi/console.fs [hunk 5]
```diff
@@ -377,7 +378,25 @@ type internal ReadLineConsole() =
...
+                    Console.Write ( x.Prompt)
+                    Console.Write( input.ToString())
+                    anchor := Anchor.Top(x.Inset)   
+            let previous = history.Previous()
+            history.Next() |> ignore
+            if previous = "" then
+                setPrompt(true)
+            else
+                setPrompt(previous.[previous.Length - 2..previous.Length - 1] = ";;")            
+
```
[the remark this answers is not in the document]
Comment (voronoipotato) [reply src/fsharp/fsi/console.fs]: removed trailing whitespace.

File: src/fsharp/fsi/console.fs [hunk 6]
```diff
@@ -377,7 +378,25 @@ type internal ReadLineConsole() =
                 if (line.Length > 0) then 
                     history.AddLast(line);
                 line;
-        
+        let clear() = 
+            current := input.Length
+            let setPrompt(prompt) =
```
[the remark this answers is not in the document]
Comment (voronoipotato) [reply src/fsharp/fsi/console.fs]: `setPrompt(prompt)` was changed to `setPrompt prompt`.

**Comment by voronoipotato** [comment]
Alright I've responded to all the feedback. The changes were done in https://github.com/dotnet/fsharp/pull/5808/commits/296f3f267f3a8977df39add7390c9fcd5a8d1d9a . Please let me know if you have any additional questions or if I've horribly misunderstood anything :) .

**Comment by cartermp** [comment]
@voronoipotato Could you also resolve conflicts?

**Comment by voronoipotato** [comment]
@cartermp  it is done. Most of the conflicts were from me unconsciously adding semicolons after spending too much time in C#.

**Comment by voronoipotato** [comment]
Linux CI failed with 

```
-> Could not download FSharpVSPowerTools.Core 2.3.
-> Couldn't get package details for package FSharpVSPowerTools.Core 2.3 on https://www.nuget.org/api/v2.
-> All possible sources are already blacklisted. 
```

**Comment by cartermp** [comment]
@mmitche we get failures like this on the Linux agents all the time:

```
Paket failed with
-> Could not download FSharpVSPowerTools.Core 2.3.
-> Couldn't get package details for package FSharpVSPowerTools.Core 2.3 on https://www.nuget.org/api/v2.
-> All possible sources are already blacklisted. 
	 - 1_https://www.nuget.org/api/v2/Packages(Id='fsharpvspowertools.core',Version='2.3')
	 - 1_https://www.nuget.org/api/v2/Packages(Id='fsharpvspowertools.core',Version='2.3.0')
	 - 1_https://www.nuget.org/api/v2/Packages(Id='FSharpVSPowerTools.Core',Version='2.3.0')
	 - 1_https://www.nuget.org/api/v2/Packages(Id='FSharpVSPowerTools.Core',Version='2.3')
	 - 2_https://www.nuget.org/api/v2/Packages?$filter=(Id eq 'FSharpVSPowerTools.Core') and (NormalizedVersion eq '2.3.0')
	 - 2_https://www.nuget.org/api/v2/Packages?$filter=(tolower(Id) eq 'fsharpvspowertools.core') and (NormalizedVersion eq '2.3.0')
	 - 2_https://www.nuget.org/api/v2/Packages?$filter=(Id eq 'FSharpVSPowerTools.Core') and (Version eq '2.3.0')
	 - 2_https://www.nuget.org/api/v2/Packages?$filter=(tolower(Id) eq 'fsharpvspowertools.core') and (Version eq '2.3')
	 - 2_https://www.nuget.org/api/v2/Packages?$filter=(tolower(Id) eq 'fsharpvspowertools.core') and (Version eq '2.3.0')
```

It's not just this package, it could be a different one at a different time. Is this something we're aware of overall?

**Comment by mmitche** [comment]
I haven't seen such an error.  @dotnet/dnceng does this ring a bell at all?

**Comment by MattGal** [comment]
@mmitche it rings a bell but I've never seen a mono SSL implementation implicated as the culprit:

```
Source 'https://www.nuget.org/api/v2' exception: System.Exception: error ---> System.Exception: Could not retrieve data from 'https://www.nuget.org/api/v2/Packages?$filter=(tolower(Id) eq 'fsharpvspowertools.core') and (Version eq '2.3.0')' ---> System.Net.Http.HttpRequestException: An error occurred while sending the request ---> System.Net.WebException: Error: TrustFailure (Authentication failed, see inner exception.) ---> System.Security.Authentication.AuthenticationException: Authentication failed, see inner exception. ---> Mono.Btls.MonoBtlsException: Ssl error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED
  at /build/mono-5.20.1.19/external/boringssl/ssl/handshake_client.c:1132
```
... the exceptions are kind of similar to when we were using desktop .NET and Nuget moved to TLS 1.2 being mandatory.  That said, that issue was 100% repro so perhaps we need to ping some mono folks if it's intermittent.

**Comment by voronoipotato** [comment]
@KevinRansom any reservations?

**Comment by cartermp** [comment]
All green, @KevinRansom for review

**Comment by cartermp** [comment]
Also adding @brettfo.

@KevinRansom is on vacation and may/may not get to things for the remainder of the month.

**Comment by KevinRansom** [comment]
@voronoipotato,  thanks for doing this mate.  I'm sorry it dropped off my radar for such a long time.  I will try to do better next time.

Anyway ... thanks

Kevin

**Comment by voronoipotato** [comment]
@KevinRansom it's cool, it was relatively low priority. Thanks for taking the time out so that it could be pulled through. 👍

**Pull request merged**